[DOCS-4556] Add Observability Pipelines Worker Aggregator Architecture doc (#16395)
DataDog/documentation: content/en/observability_pipelines/production_deployment_overview/aggregator_architecture.md

- **Improved security.** Edge nodes can be locked down, aggregator nodes can be hardened.
- **Optimized delivery.** Data can be streamlined for optimal delivery to destinations.
- **Service reliability.** Doing less on the edge reduces agent-induced incidents.
- **Reduced lock-in.** Collect data from any source and send it to any destination.
- **Global aggregation.** Data can be aggregated and analyzed across multiple nodes.

#### Disadvantages

- An entire aggregator outage can be a single point of failure for your observability data pipeline. Therefore, use [high availability](#high-availability) tactics to help mitigate this issue.

 ## Requirements

| Type              | Minimum Value                                         			|
| ----------------- | ----------------------------------------------------------------- |
| CPU Cores         | ≥ 2 vCPUs (see [CPU sizing](#cpu-sizing))             			|
| CPU Architectures | X86_64, AMD64, ARM64, ARMHF, ARMv7                   		 		|
| Memory            | ≥ 2 GiB per vCPU (see [memory sizing](#memory-sizing))			|
| Disk              | ≥ 1 Gib, more for disk buffers (see [disk sizing](#disk-sizing))  |

 ## Quickstart

>  See [Setup][link to come] for installing the Observability Pipelines Worker.

Author: I'll add the link to the Setup page once that is available.

##### Load balancer hot spots

Load balancing hot spots occur when one or more Observability Pipelines Worker instance receives disproportionate traffic. Hot spots usually happen due to one of two reasons:

1. A substantial amount of traffic is being sent over a single connection.
2. Traffic in one availability zone is much higher than in the others.

In these cases, the following respective mitigation tactics are recommended:

1. Split large connections into multiple connections. Most clients allow connection concurrency and pooling that distributes data over multiple connections. This tactic allows your load balancer to distribute the connection across multiple Observability Pipelines Worker instances. If your client does not support this, consider the unified architecture, where Observability Pipelines Worker can be additionally deployed to the edge.
2. Enable cross-zone load balancing on your load balancer. Cross-zone balancing balances all availability zone traffic across all Observability Pipelines Worker instances.

### Vertical scaling

> Observability Pipelines Worker's concurrency model automatically scales to take advantage of all vCPUs. There are no concurrency settings or configuration changes required. When vertically scaling, Datadog recommends capping an instance's size to process no more than 50% of your total volume to force deploy at least two Observability Pipelines Worker instances for high availability.

Author: ```suggestion
Observability Pipelines Worker's concurrency model automatically scales to take advantage of all vCPUs. There are no concurrency settings or configuration changes required. When vertically scaling, Datadog recommends capping an instance's size to process no more than 50% of your total volume and deploying at least two Observability Pipelines Worker instances for high availability.
```

- More [advanced configurations](#advanced-configurations) for deploying multiple aggregators, pub-sub systems, and global aggregation

### Comparison

Compared to the Observability Pipelines Worker agent architecture, the aggregator architecture has the following advantages and disadvantages.

#### Advantages

- **Easy to set up.** Deploy Observability Pipelines Worker like any other service.
- **Easy to adopt.** Data sources can be onboarded incrementally, one source at a time.
- **Reduced footprint.** This architecture touches fewer nodes.
- **High availability.** Multi-node failover mitigates failure scenarios.
- **High durability.** Centralized routing allows for sophisticated durability strategies.
- **Improved security.** Edge nodes can be locked down, aggregator nodes can be hardened.
- **Optimized delivery.** Data can be streamlined for optimal delivery to destinations.
- **Service reliability.** Doing less on the edge reduces agent-induced incidents.
- **Reduced lock-in.** Collect data from any source and send it to any destination.
- **Global aggregation.** Data can be aggregated and analyzed across multiple nodes.

#### Disadvantages

> - An entire aggregator outage can be a single point of failure for your observability data pipeline. Therefore, use [high availability](#high-availability) tactics to help mitigate this issue.

Author: I agree. @barieom, can we remove this section? From what I understand, we will also be getting rid of the Agent role and focus on the Aggregator, so it seems like we won't need this section anyway.

> - An entire aggregator outage can be a single point of failure for your observability data pipeline. Therefore, use [high availability](#high-availability) tactics to help mitigate this issue.

Author: Great, thanks, removing the section.

##### Load balancer hot spots

Load balancing hot spots occur when one or more Observability Pipelines Worker instances receives disproportionate traffic. Hot spots usually happen due to one of two reasons:

1. A substantial amount of traffic is being sent over a single connection.
2. Traffic in one availability zone is much higher than in the others.

In these cases, the following respective mitigation tactics are recommended:

1. Split large connections into multiple connections. Most clients allow connection concurrency/pooling that distributes data over multiple connections. This tactic allows your load balancer to distribute the connection across multiple Observability Pipelines Worker instances. If your client does not support this, consider the unified architecture, where Observability Pipelines Worker can be additionally deployed to the edge.
2. Enable cross-zone load balancing on your load balancer. Cross-zone balancing balances all availability zone traffic across all Observability Pipelines Worker instances.

### Vertical scaling

> Observability Pipelines Worker's concurrency model automatically scales to take advantage of all vCPUs. There are no concurrency settings or configuration changes required. When vertically scaling, Datadog recommends capping an instance's size to process no more than 50% of your total volume to force deploy at least two Observability Pipelines Worker instances for high availability.

Author: @barieom

> Observability Pipelines Worker's concurrency model automatically scales to take advantage of all vCPUs. There are no concurrency settings or configuration changes required. When vertically scaling, Datadog recommends capping an instance's size to process no more than 50% of your total volume to force deploy at least two Observability Pipelines Worker instances for high availability.

Author: That's a lot clearer, will make the change.

### Autoscaling

Autoscaling should be based on average CPU utilization. For the vast majority of workloads, Observability Pipelines Worker is CPU constrained. CPU utilization is the strongest signal for autoscaling since it does not produce false positives. The following settings are recommended, adjust as necessary:

- Average CPU with a 85% utilization target.
- A five minute stabilization period for scaling up and down.

## Networking

### Network topology

#### Network boundaries

> Most users have complex production environments with many network boundaries, including multiple clouds, regions, VPCs, and clusters. It can get complicated when determining where Observability Pipelines Worker fits within those boundaries. Therefore, **starting with one Observability Pipelines Worker aggregator per region is recommended**, even if you have multiple accounts, VPCs, and clusters. This boundary is the broadest networking granularity that avoids sending data over the public internet. If you have multiple clusters, deploy Observability Pipelines Worker into your utility or tools cluster, or pick a cluster that is most appropriate for shared services.

Author: Thanks, Jorie, for letting me know! I'll go with your suggestion.

The Observability Pipelines Worker can have problems, such as failing to parse a log, when trying to process malformed data. There are two ways to mitigate this issue:

1. **Direct archiving**: Route data directly from your sources to your archive. This ensures that data makes it to your archive without risk of being dropped. In addition, this data can be replayed after correcting the processing error.

2. **Failed event routing**: The Observability Pipelines Worker offers failed event routing for users who wish to archive processed data, such as structured and enriched data. Certain Observability Pipelines Worker transforms come with a dropped output that can be connected to a sink for durability and replay.

#### Which strategy is best?

If durability is the most important criteria, use the direct archiving method because it addresses data loss scenarios. Use the failed event routing method, also commonly referred to as a data lake, if you prefer to analyze data in your archive. It has the advantage of using your archive/data lake for long-term analysis. Datadog [Log Archives][2] and AWS Athena are examples of archive storage solutions.

### Handling destination failures

> Finally, destination failures refer to the total failure of a downstream destination (for example Elasticsearch). Data loss can be mitigated for issues with the downstream destination by using disk buffers large enough to sustain the outage time. This allows data to durably buffer while the service is down and then drain when the service comes back up. For this reason, disk buffers large enough to hold at least one hour's worth of data are recommended. See [disk sizing](#disk-sizing) for more details.

Reviewer: ```suggestion
Destination failures refer to the total failure of a downstream destination (for example, Elasticsearch). Data loss can be mitigated for issues with the downstream destination by using disk buffers large enough to sustain the outage time. This allows data to durably buffer while the service is down and then drain when the service comes back up. For this reason, disk buffers large enough to hold at least one hour's worth of data are recommended. See [disk sizing](#disk-sizing) for more details.
```

## High availability

High availability refers to Observability Pipelines Worker remaining available if there are any system issues.

To achieve high durability:

1. Deploy at least two Observability Pipelines Worker instances in each Availability Zone.
2. Deploy Observability Pipelines Worker in at least two Availability Zones.
3. Front your Observability Pipelines Worker instances with a load balancer that balances traffic across Observability Pipelines Worker instances. See the horizontal scaling section for more information.

### Mitigating failure scenarios

#### Handling Observability Pipelines Worker process issues

> To mitigate a system process issue, distribute Observability Pipelines Worker across multiple nodes and front them with a network load balancer that can redirect traffic to another Observability Pipelines Worker instance as needed. In addition, platform-level automated self-healing should eventually restart the process or replace the node.

Reviewer: ```suggestion
To mitigate a system process issue, distribute the Observability Pipelines Worker across multiple nodes and front them with a network load balancer that can redirect traffic to another Observability Pipelines Worker instance as needed. In addition, platform-level automated self-healing should eventually restart the process or replace the node.
```

#### Mitigating node failures

To mitigate node issues, distribute the Observability Pipelines Worker across multiple nodes and front them with a network load balancer that can redirect traffic to another Observability Pipelines Worker node. In addition, platform-level automated self-healing should eventually replace the node.

> 

Reviewer: These two paragraphs are almost identical. Can we condense them to a single paragraph?

## Disaster recovery

### Internal disaster recovery

Observability Pipelines Worker is an infrastructure-level tool designed to route internal observability data. It implements a shared-nothing architecture and does not manage state that should be replicated or transferred to a disaster recovery (DR) site. Therefore, if your entire region fails, Observability Pipelines Worker would fail with it. Therefore, you should install the Observability Pipelines Worker in your DR site as part of your broader DR plan.

### External disaster recovery

If you're using a managed destination, such as Datadog, Observability Pipelines Worker can facilitate automatic routing data to your Datadog DR site using Observability Pipelines Worker's circuit breaker feature.

## Advanced configurations

### Multiple aggregator deployments

> As covered in [network boundaries](#network-boundaries), Datadog recommends to start with one Observability Pipelines Worker aggregator per region. This is to prevent overcomplicating your initial deployment of Observability Pipelines Worker, but there are circumstances where starting with multiple deployments makes sense:

Reviewer: ```suggestion
As covered in [network boundaries](#network-boundaries), Datadog recommends to start with one Observability Pipelines Worker aggregator per region. This is to prevent overcomplicating your initial deployment of Observability Pipelines Worker, but there are circumstances where starting with multiple deployments is ideal:
```

1. **Prevent sending data over the public internet.** If you have multiple clouds and regions, deploy the Observability Pipelines Worker aggregator in each of them to prevent sending large amounts of data over the internet. Your Observability Pipelines Worker aggregator should receive internal data and serve as the single point of egress for your network.

> 2. **Independent management.** You have teams that can operate and manage a Observability Pipelines Worker aggregator independently for their use case. For example, your Data Science team may be responsible for operating their own infrastructure and has the means to independently operate their own Observability Pipelines Worker aggregator.

Reviewer: ```suggestion
2. **Independent management.** You have teams that can operate and manage an Observability Pipelines Worker aggregator independently for their use case. For example, your Data Science team may be responsible for operating their own infrastructure and has the means to independently operate their own Observability Pipelines Worker aggregator.
```

### Multiple cloud accounts

> Many users have multiple cloud accounts with VPCs and clusters inside. Datadog still recommends in this case to deploy one Observability Pipelines Worker aggregator per region. Deploy Observability Pipelines Worker into your utility/tools cluster and configure all your cloud accounts to send data to this cluster. See [network boundaries](#network-boundaries) for more information.

Reviewer: ```suggestion
Many users have multiple cloud accounts with VPCs and clusters inside. Datadog still recommends in this case to deploy one Observability Pipelines Worker aggregator per region. Deploy Observability Pipelines Worker into your utility or tools cluster and configure all your cloud accounts to send data to this cluster. See [network boundaries](#network-boundaries) for more information.
```

> ### Pub-sub systems

Reviewer: ```suggestion
### Publish-subscribe (pub-sub) systems
```

> Using a pub-sub system, like Kafka, is not required to make this architecture highly available or highly durable (see the high durability and high availability sections), but they do offer the following advantages:

Reviewer: ```suggestion
Using a pub-sub system such as Kafka is not required to make your architecture highly available or highly durable (see the high durability and high availability sections), but they do offer the following advantages:
```

> 1. **Improved reliability.** Pub-sub systems are designed to be highly reliable and durable systems that change infrequently. They are especially reliable if you are using a managed option. The Observability Pipelines Worker is likely to change more often by nature of its purpose. Insulate Observability Pipelines Worker downtime behind a pub-sub system to increase availability from the perception of your clients and make recovery simpler.

Reviewer: ```suggestion
1. **Improved reliability.** Pub-sub systems are designed to be highly reliable and durable systems that change infrequently. They are especially reliable if you are using a managed option. The Observability Pipelines Worker is likely to change more often based on its purpose. Isolate Observability Pipelines Worker downtime behind a pub-sub system to increase availability from the perception of your clients and make recovery simpler.
```

2. **Load balancer not required.** Pub-sub systems eliminate the need for a load balancer. Your pub-sub system handles the coordination of consumers, making it easy to scale Observability Pipelines Worker horizontally.

#### Pub-sub partitioning

> Partitioning, or "topics" in Kafka terminology, refers to separating data in your pub-sub systems. Datadog strongly recommends partitioning along data origin lines, such as the service or host that generated the data.

Reviewer: ```suggestion
Partitioning, or "topics" in Kafka terminology, refers to separating data in your pub-sub systems. You should partition along data origin lines, such as the service or host that generated the data.
```

#### Pub-sub configuration

> When using a pub-sub system, the following configuration changes for Observability Pipelines Worker are recommended:

Reviewer: ```suggestion
When using a pub-sub system, Datadog recommends the following configuration changes for Observability Pipelines Worker:
```

- **Enable end-to-end acknowledgements for all sinks.** This setting ensures that the pub-sub checkpoint is not advanced until data is successfully written.
- **Use memory buffers.** There is no need to use Observability Pipelines Worker's disk buffers when it sits behind a pub-sub system. Your pub-sub system is designed for long-term buffering with high durability. Observability Pipelines Worker should only be responsible for reading, processing, and routing the data (not durability).

### Global aggregation

> This section provides recommendations for performing global calculations for legacy destinations. Modern destinations already support global calculations. For example, Datadog supports distributions (DDSketch) that solve global observations of your metrics data.

Reviewer: ```suggestion
This section provides recommendations for performing global calculations for legacy destinations. Modern destinations already support global calculations. For example, Datadog supports distributions (such as DDSketch) that solve global observations of your metrics data.
```

Global aggregation refers to the ability to aggregate data for an entire region. For example, computing global quantiles for CPU load averages. To achieve this, a single Observability Pipelines Worker instance must have access to every node's CPU load average statistics. This is not possible with horizontal scaling; each individual Observability Pipelines Worker instance only has access to a slice of the overall data. Therefore, aggregation should be tiered.

> The structure is that tier 2 aggregators receive an aggregated sub-stream of the overall data from tier 1 aggregators. This allows a single instance to get a global view without processing the entire stream and introducing a single point of failure.

Reviewer: ```suggestion
The structure is that tier two aggregators receive an aggregated sub-stream of the overall data from tier one aggregators. This allows a single instance to get a global view without processing the entire stream and introducing a single point of failure.
```

#### Recommendations

- Limit global aggregation to tasks that can reduce data, such as computing global histograms. Never send all data to your global aggregators.

> - Continue to use your local aggregators to process and deliver most data so that you do notintroduce a single point of failure.

Reviewer: ```suggestion
- Continue to use your local aggregators to process and deliver most data so that you do not introduce a single point of failure.
```
